onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -156,15 +162,21 @@ static common::Status DeserializeTensorProto(const Env& env, const std::basic_st
 
     // deserialize to CPU first for non-CPU allocator, then copy
     std::unique_ptr<Tensor> p_deserialize_tensor;
-    if (use_device_allocator_for_initializers) {
-      void* tensor_buffer = nullptr;
-      ORT_RETURN_IF_ERROR(AllocateBufferUsingDeviceAllocatorFromShapeAndType(tensor_shape, type, default_cpu_alloc, tensor_buffer));
-      p_deserialize_tensor = std::make_unique<Tensor>(type, tensor_shape, tensor_buffer, default_cpu_alloc);
+
+    // for external initializer we will use mmap so don't need to allocate memory in advance
+    if (!utils::HasExternalData(tensor_proto)) {
```

Comment:
Didn't realize that this was repeated here. This function ```DeserializeTensorProto``` is getting really hard to read. Potential for cleanup.

Reply:
Yes, I have discussion with Yulong today about this part as well, I will re-organize this part.